diff --git a/model/CMakeLists.txt b/model/CMakeLists.txt
index 1849d68..0cfabbb 100644
--- a/model/CMakeLists.txt
+++ b/model/CMakeLists.txt
@@ -234,6 +234,24 @@ foreach (xlen IN ITEMS 32 64)
                     set(coverage_args)
                 endif()
 
+                add_custom_command(
+                    DEPENDS ${sail_srcs}
+                    OUTPUT /tmp/raw-${arch}.rs
+                    VERBATIM
+                    COMMENT "Building Rust code from Sail model (${arch})"
+                    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+                    COMMAND
+                        ${SAIL_BIN}
+                        # Output file (without extension).
+                        -o ${CMAKE_SOURCE_DIR}/../../rv64/src/raw.rs
+                        -plugin ${CMAKE_SOURCE_DIR}/../../_build/default/sail_rust_backend/sail_plugin_rust.cmxs
+                        -rust
+                        # Input files.
+                        ${sail_srcs}
+                )
+
+                add_custom_target(generated_rust_${arch} DEPENDS /tmp/raw-${arch}.rs)
+
                 add_custom_command(
                     DEPENDS ${sail_srcs}
                     OUTPUT ${c_model} ${branch_info_file}
diff --git a/model/riscv_insts_vext_mask.sail b/model/riscv_insts_vext_mask.sail
index d9a4847..0ec92c9 100644
--- a/model/riscv_insts_vext_mask.sail
+++ b/model/riscv_insts_vext_mask.sail
@@ -32,7 +32,7 @@ mapping clause encdec = MMTYPE(funct6, vs2, vs1, vd)
 function clause execute(MMTYPE(funct6, vs2, vs1, vd)) = {
   let SEW      = get_sew();
   let LMUL_pow = get_lmul_pow();
-  let num_elem = VLEN;
+  let num_elem = get_vlen();
 
   if illegal_vd_unmasked() then return Illegal_Instruction();
 
@@ -93,7 +93,7 @@ mapping clause encdec = VCPOP_M(vm, vs2, rd)
 function clause execute(VCPOP_M(vm, vs2, rd)) = {
   let SEW      = get_sew();
   let LMUL_pow = get_lmul_pow();
-  let num_elem = VLEN;
+  let num_elem = get_vlen();
 
   if illegal_vd_unmasked() | not(assert_vstart(0)) then return Illegal_Instruction();
 
@@ -131,7 +131,7 @@ mapping clause encdec = VFIRST_M(vm, vs2, rd)
 function clause execute(VFIRST_M(vm, vs2, rd)) = {
   let SEW      = get_sew();
   let LMUL_pow = get_lmul_pow();
-  let num_elem = VLEN;
+  let num_elem = get_vlen();
 
   if illegal_vd_unmasked() | not(assert_vstart(0)) then return Illegal_Instruction();
 
@@ -171,7 +171,7 @@ mapping clause encdec = VMSBF_M(vm, vs2, vd)
 function clause execute(VMSBF_M(vm, vs2, vd)) = {
   let SEW      = get_sew();
   let LMUL_pow = get_lmul_pow();
-  let num_elem = VLEN;
+  let num_elem = get_vlen();
 
   if illegal_normal(vd, vm) | not(assert_vstart(0)) | vd == vs2
   then return Illegal_Instruction();
@@ -215,7 +215,7 @@ mapping clause encdec = VMSIF_M(vm, vs2, vd)
 function clause execute(VMSIF_M(vm, vs2, vd)) = {
   let SEW      = get_sew();
   let LMUL_pow = get_lmul_pow();
-  let num_elem = VLEN;
+  let num_elem = get_vlen();
 
   if illegal_normal(vd, vm) | not(assert_vstart(0)) | vd == vs2
   then return Illegal_Instruction();
@@ -259,7 +259,7 @@ mapping clause encdec = VMSOF_M(vm, vs2, vd)
 function clause execute(VMSOF_M(vm, vs2, vd)) = {
   let SEW      = get_sew();
   let LMUL_pow = get_lmul_pow();
-  let num_elem = VLEN;
+  let num_elem = get_vlen();
 
   if illegal_normal(vd, vm) | not(assert_vstart(0)) | vd == vs2
   then return Illegal_Instruction();
diff --git a/model/riscv_insts_vext_mem.sail b/model/riscv_insts_vext_mem.sail
index ba5c838..dc1e0ea 100644
--- a/model/riscv_insts_vext_mem.sail
+++ b/model/riscv_insts_vext_mem.sail
@@ -619,7 +619,7 @@ function process_vlre (nf, vd, load_width_bytes, rs1, elem_per_reg) = {
 function clause execute(VLRETYPE(nf, rs1, width, vd)) = {
   let load_width_bytes = vlewidth_bytesnumber(width);
   let EEW = load_width_bytes * 8;
-  let elem_per_reg : int = VLEN / EEW;
+  let elem_per_reg : int = get_vlen() / EEW;
   let nf_int = nfields_int(nf);
 
   assert(elem_per_reg >= 0);
@@ -687,7 +687,7 @@ function process_vsre (nf, load_width_bytes, rs1, vs3, elem_per_reg) = {
 function clause execute(VSRETYPE(nf, rs1, vs3)) = {
   let load_width_bytes = 1;
   let EEW = 8;
-  let elem_per_reg : int = VLEN / EEW;
+  let elem_per_reg : int = get_vlen() / EEW;
   let nf_int = nfields_int(nf);
 
   assert(elem_per_reg >= 0);
diff --git a/model/riscv_pmp_regs.sail b/model/riscv_pmp_regs.sail
index 0eacf8c..a518ca0 100644
--- a/model/riscv_pmp_regs.sail
+++ b/model/riscv_pmp_regs.sail
@@ -231,7 +231,7 @@ mapping clause csr_name_map = 0x3EE  <-> "pmpaddr62"
 mapping clause csr_name_map = 0x3EF  <-> "pmpaddr63"
 
 // pmpcfgN
-function clause is_CSR_defined(0x3A) @ idx : bits(4) = sys_pmp_count() > unsigned(idx) & (idx[0] == bitzero | xlen == 32)
+function clause is_CSR_defined(0x3A) @ idx : bits(4) = sys_pmp_count() > 4 * unsigned(idx) & (idx[0] == bitzero | xlen == 32)
 function clause read_CSR(0x3A @ idx : bits(4) if idx[0] == bitzero | xlen == 32) = pmpReadCfgReg(unsigned(idx))
 function clause write_CSR((0x3A @ idx : bits(4), value) if idx[0] == bitzero | xlen == 32) = {
   let idx = unsigned(idx);
diff --git a/model/riscv_vext_regs.sail b/model/riscv_vext_regs.sail
index 5866d28..7fc7c30 100644
--- a/model/riscv_vext_regs.sail
+++ b/model/riscv_vext_regs.sail
@@ -173,7 +173,7 @@ function wV (Vregno(r) : vregno, v : vregtype) -> unit = {
 
   dirty_v_context();
 
-  assert(0 < VLEN & VLEN <= sizeof(vlenmax));
+  assert(0 < get_vlen() & get_vlen() <= sizeof(vlenmax));
   vreg_write_callback(vregno_to_vregidx(Vregno(r)), v);
 }
 
@@ -242,7 +242,7 @@ function get_num_elem(LMUL_pow, SEW) = {
   let LMUL_pow_reg = if LMUL_pow < 0 then 0 else LMUL_pow;
   /* Ignore lmul < 1 so that the entire vreg is read, allowing all masking to
    * be handled in init_masked_result */
-  let num_elem = 2 ^ LMUL_pow_reg * VLEN / SEW;
+  let num_elem = 2 ^ LMUL_pow_reg * get_vlen() / SEW;
   assert(num_elem > 0);
   num_elem
 }
@@ -294,7 +294,7 @@ function read_vreg(num_elem, SEW, LMUL_pow, vrid) = {
     if LMUL_pow < 0 then {
       result = read_single_vreg('n, SEW, vrid);
     } else {
-      let 'num_elem_single : int = VLEN / SEW;
+      let 'num_elem_single : int = get_vlen() / SEW;
       assert('num_elem_single >= 0);
       foreach (i_lmul from 0 to (2 ^ LMUL_pow_reg - 1)) {
         let r_start_i : int = i_lmul * 'num_elem_single;
@@ -317,8 +317,8 @@ function read_vreg(num_elem, SEW, LMUL_pow, vrid) = {
 /* Single element reading operation */
 val read_single_element : forall 'm 'x, 8 <= 'm <= 128. (int('m), int('x), vregidx) -> bits('m)
 function read_single_element(EEW, index, vrid) = {
-  assert(VLEN >= EEW);
-  let 'elem_per_reg : int = VLEN / EEW;
+  assert(get_vlen() >= EEW);
+  let 'elem_per_reg : int = get_vlen() / EEW;
   assert('elem_per_reg >= 0);
   let real_vrid  : vregidx = vregidx_offset(vrid, to_bits(5, index / 'elem_per_reg));
   let real_index : int    = index % 'elem_per_reg;
@@ -332,7 +332,7 @@ val write_vreg : forall 'n 'm 'p, 'n >= 0 & 'm >= 0. (int('n), int('m), int('p),
 function write_vreg(num_elem, SEW, LMUL_pow, vrid, vec) = {
   let LMUL_pow_reg = if LMUL_pow < 0 then 0 else LMUL_pow;
 
-  let 'num_elem_single  : int = VLEN / SEW;
+  let 'num_elem_single  : int = get_vlen() / SEW;
   assert('num_elem_single >= 0);
   foreach (i_lmul from 0 to (2 ^ LMUL_pow_reg - 1)) {
     var single_vec : vector('num_elem_single, bits('m)) = vector_init(zeros());
@@ -352,7 +352,7 @@ function write_vreg(num_elem, SEW, LMUL_pow, vrid, vec) = {
 /* Single element writing operation */
 val write_single_element : forall 'm 'x, 8 <= 'm <= 128. (int('m), int('x), vregidx, bits('m)) -> unit
 function write_single_element(EEW, index, vrid, value) = {
-  let 'elem_per_reg : int = VLEN / EEW;
+  let 'elem_per_reg : int = get_vlen() / EEW;
   assert('elem_per_reg >= 0);
   let real_vrid  : vregidx = vregidx_offset(vrid, to_bits(5, index / 'elem_per_reg));
   let real_index : int    = index % 'elem_per_reg;
@@ -409,6 +409,7 @@ function read_vmask_carry(num_elem, vm, vrid) = {
 /* Mask register writing operation with num_elem as max(VLMAX,VLEN/SEW)) */
 val write_vmask : forall 'n, 'n > 0. (int('n), vregidx, bits('n)) -> unit
 function write_vmask(num_elem, vrid, v) = {
+  let VLEN: range(8, 65536) = get_vlen();
   assert(0 < VLEN & VLEN <= sizeof(vlenmax));
   assert(0 < num_elem & num_elem <= VLEN);
   let vreg_val : vregtype = V(vrid);
diff --git a/model/riscv_vlen.sail b/model/riscv_vlen.sail
index 2367b72..8fd41ce 100644
--- a/model/riscv_vlen.sail
+++ b/model/riscv_vlen.sail
@@ -17,4 +17,4 @@ function get_vlen_pow() -> range(3, 16) = config extensions.V.vlen_exp
 
 type vlenmax : Int = 65536
 
-let VLEN = 2 ^ get_vlen_pow()
+function get_vlen() -> range(8, 65536) = 2 ^ get_vlen_pow()
diff --git a/model/riscv_zvk_utils.sail b/model/riscv_zvk_utils.sail
index 38ca001..2e59def 100644
--- a/model/riscv_zvk_utils.sail
+++ b/model/riscv_zvk_utils.sail
@@ -14,7 +14,15 @@ function zvk_valid_reg_overlap(rs, rd, emul_pow) = {
   (rs_int + reg_group_size <= rd_int) | (rd_int + reg_group_size <= rs_int)
 }
 
-function zvk_check_encdec(EGW: int, EGS: int) -> bool = (unsigned(vl) % EGS == 0) & (unsigned(vstart) % EGS == 0) & (2 ^ get_lmul_pow() * VLEN) >= EGW
+function zvk_check_encdec(EGW: int, EGS: int) -> bool = {
+  let LMUL_pow = get_lmul_pow();
+  let LMUL_times_VLEN : int =
+    if LMUL_pow < 0
+    then get_vlen() / (2 ^ abs_int(LMUL_pow))
+    else (2 ^ LMUL_pow) * get_vlen()
+  in
+  (unsigned(vl) % EGS == 0) & (unsigned(vstart) % EGS == 0) & LMUL_times_VLEN >= EGW
+}
 
 /*
  * Utility functions for Zvknh[ab]
